File: ros2-for-beginners/ros2_ws/src/my_py_pkg/my_py_pkg/smartphone.py
#!usr/bin/python3

import rclpy
from rclpy.node import Node
from example_interfaces.msg import String


class SmartPhoneNode(Node):
    def __init__(self) -> None:
        super().__init__('smartphone')

        '''
        create a subscription object 
        using create_subscription method of Node class
        takes in 4 arguements
            1. message type 
            2. name of the topic to which the node want to subscribe 
            3. callback function --> every time a message is published to a topic this callback function is invoked
                it will contain the received message('msg') as parameter
            4. message queue size 
        '''
        self.subscriber_ = self.create_subscription(
            String, "robot_news", self.callback_robot_news, 10)

        # log
        self.get_logger().info('smartphone node started')

    # create a callback function for handling what to do with the message
    def callback_robot_news(self, msg):
        self.get_logger().info(f"recieved message : {msg.data}")


def main():
    rclpy.init()
    node = SmartPhoneNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2-for-beginners/ros2_ws/src/my_cpp_pkg/src/smartphone.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

class SmartPhoneNode : public rclcpp::Node
{

public:
    SmartPhoneNode() : Node("smartphone")
    {

        /*
        initialize subscriber
            the function takes in 3 parameters they are
                1. topic name
                2. queue size
                3. callback --> additional code required since the callback function takes in parameters.
        */
        subscriber_ = this->create_subscription<example_interfaces::msg::String>("robot_news", 10,
                                                                                 std::bind(&SmartPhoneNode::callbackRobotNews, this, std::placeholders::_1));
        RCLCPP_INFO(this->get_logger(), "cpp smartphone node has been started");
    }

private:
    // call back for handling subscription
    void callbackRobotNews(const example_interfaces::msg::String::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "%s", msg->data.c_str());
    }

    // declaration
    rclcpp::Subscription<example_interfaces::msg::String>::SharedPtr subscriber_;

};

int
main(int argc, char **argv)
{
    rclcpp::init(argc,argv);
    auto node = std::make_shared<SmartPhoneNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}